Source/Diagnostics/ComputeDiagFunctors/TemperatureFunctor.cpp

```diff
@@ -0,0 +1,165 @@
+                const amrex::ParticleReal uz = p.rdata(PIdx::uz);
+                amrex::Gpu::Atomic::AddNoRet(&out_array(ii, jj, kk, 0), (amrex::Real)(w));
+                amrex::Gpu::Atomic::AddNoRet(&out_array(ii, jj, kk, 1), (amrex::Real)(w*ux));
+                amrex::Gpu::Atomic::AddNoRet(&out_array(ii, jj, kk, 2), (amrex::Real)(w*uy));
+                amrex::Gpu::Atomic::AddNoRet(&out_array(ii, jj, kk, 3), (amrex::Real)(w*uz));
+            });
+
+    // Divide value by number of particles for average
+    for (amrex::MFIter mfi(sum_mf, amrex::TilingIfNotGPU()); mfi.isValid(); ++mfi)
+    {
+        const amrex::Box& box = mfi.tilebox();
+        amrex::Array4<amrex::Real> const& out_array = sum_mf.array(mfi);
+        amrex::ParallelFor(box,
+                [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    if (out_array(i,j,k,0) > 0) {
+                        const amrex::Real invsum = 1._rt/out_array(i,j,k,0);
+                        out_array(i,j,k,1) *= invsum;
+                        out_array(i,j,k,2) *= invsum;
+                        out_array(i,j,k,3) *= invsum;
+                    }
+                });
+    }
+
+    // Calculate the sum of the squares, subtracting the averages
+    // These loops must be written out since ParticleToMesh always zeros out the mf.
+    const auto plo = pc.Geom(m_lev).ProbLoArray();
+    const auto dxi = pc.Geom(m_lev).InvCellSizeArray();
+    for (WarpXParIter pti(pc, m_lev); pti.isValid(); ++pti)
+    {
+        const long np = pti.numParticles();
+        amrex::ParticleReal* wp = pti.GetAttribs(PIdx::w).dataPtr();
+        amrex::ParticleReal* uxp = pti.GetAttribs(PIdx::ux).dataPtr();
+        amrex::ParticleReal* uyp = pti.GetAttribs(PIdx::uy).dataPtr();
+        amrex::ParticleReal* uzp = pti.GetAttribs(PIdx::uz).dataPtr();
+
+        auto const GetPosition = GetParticlePosition<PIdx>(pti);
+
+        amrex::Array4<amrex::Real> const& out_array = sum_mf.array(pti);
+
+        amrex::ParallelFor(np,
+            [=] AMREX_GPU_DEVICE (long ip) {
+                // --- Get particle quantities
+                amrex::ParticleReal xp, yp, zp;
+                GetPosition.AsStored(ip, xp, yp, zp);
+
+                // Get position in AMReX convention to calculate corresponding index.
+                int ii = 0, jj = 0, kk = 0;
+                const amrex::Real lx = (xp - plo[0]) * dxi[0];
+                ii = static_cast<int>(amrex::Math::floor(lx));
+#if defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+                const amrex::Real lz = (zp - plo[1]) * dxi[1];
+                jj = static_cast<int>(amrex::Math::floor(lz));
+#elif defined(WARPX_DIM_3D)
+                const amrex::Real ly = (yp - plo[1]) * dxi[1];
+                jj = static_cast<int>(amrex::Math::floor(ly));
+                const amrex::Real lz = (zp - plo[2]) * dxi[2];
+                kk = static_cast<int>(amrex::Math::floor(lz));
+#endif
```

Comment:
It would be useful to have a function (maybe in `ParticleUtils.cpp`) that returns the cell indices for a given particle. That would simplify the code here and from lines 60-71.

I'll make an issue out of this comment for follow-up.

Reply:
I would agree. This routine would be useful elsewhere, for example in the deposition and gather routines.

Thanks for merging!